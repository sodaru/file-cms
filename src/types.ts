export type AttributeFilter = string[];

export type Content = {
  type: string;
  slug: string;
  content: string;
} & Record<string, unknown>;

export type GetContent = (
  type: string,
  slug: string,
  attributes?: AttributeFilter
) => Promise<Partial<Content>>;

export type FilterProvider = (data: unknown, arg: unknown) => boolean;

export type GeneralExpression = Record<string, unknown>;
export type EqExpression = { eq: string | number | boolean | null };
export type StartsWithExpression = { startsWith: string };
export type EndsWithExpression = { endsWith: string };
export type ContainsExpression = { contains: string };
export type GtExpression = { gt: number };
export type LtExpression = { lt: number };
export type GteExpression = { gte: number };
export type LteExpression = { lte: number };
export type BetweenExpression = { between: [number, number] };
export type BetweenInclusiveExpression = { betweenInclusive: [number, number] };
export type HasExpression = { has: string | number | boolean | null };

export type StringExpression =
  | string
  | EqExpression
  | StartsWithExpression
  | EndsWithExpression
  | ContainsExpression
  | GeneralExpression;

export type TypeExpression = StringExpression;

export type SlugExpression = StringExpression;

export type ContentExpression = StringExpression;

export type AnyExpression =
  | string
  | number
  | boolean
  | null
  | GeneralExpression
  | EqExpression
  | StartsWithExpression
  | EndsWithExpression
  | ContainsExpression
  | GtExpression
  | LtExpression
  | GteExpression
  | LteExpression
  | BetweenExpression
  | HasExpression;

export type ListContent = (
  filters?: {
    type?: TypeExpression;
    slug?: SlugExpression;
    content?: ContentExpression;
  } & Record<string, AnyExpression>,
  attributes?: AttributeFilter,
  customFilterProviders?: Record<string, FilterProvider>
) => Promise<Partial<Content>[]>;
