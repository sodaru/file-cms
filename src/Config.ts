import { NoRootDirSetError } from "./errors/NoRootDirSetError";
import { UnknownFilterProviderError } from "./errors/UnknownFilterProviderError";
import { defaultProviders } from "./internal/filterProviders";
import { FilterProvider } from "./types";

export class Config {
  private static rootDir: string;
  private static filterProviders: Record<string, FilterProvider> = {
    ...defaultProviders
  };

  static setRootDir(rootDir: string) {
    this.rootDir = rootDir;
  }

  static getRootDir() {
    if (typeof this.rootDir == "string") {
      return this.rootDir;
    }
    throw new NoRootDirSetError();
  }

  static setFilterProvider(name: string, provider: FilterProvider) {
    this.filterProviders[name] = provider;
  }

  static getFilterProvider(name: string) {
    if (this.filterProviders[name] === undefined) {
      throw new UnknownFilterProviderError(name);
    }
    return this.filterProviders[name];
  }
}
