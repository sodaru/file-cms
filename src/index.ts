export * from "./types";
export * from "./Config";
export * from "./getContent";
export * from "./listContent";
export * from "./getFile";
