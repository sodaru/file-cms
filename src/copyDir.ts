import { dirname, join, relative } from "path";
import { Config } from "./Config";
import { copyFile, mkdir, readdir, stat } from "fs/promises";

export const copyDir = async (sourceDir: string, targetDir: string) => {
  const rootDir = Config.getRootDir();
  const fullSourceDir = join(rootDir, sourceDir);
  const dirsToCopy = [fullSourceDir];
  while (dirsToCopy.length > 0) {
    const dirToCopy = dirsToCopy.shift();
    const files = await readdir(dirToCopy);
    await Promise.all(
      files.map(async file => {
        const fileOrDirPath = join(dirToCopy, file);
        const stats = await stat(fileOrDirPath);
        if (stats.isDirectory()) {
          dirsToCopy.push(fileOrDirPath);
        } else {
          const targetPath = join(
            targetDir,
            relative(fullSourceDir, fileOrDirPath)
          );
          const targetDirname = dirname(targetPath);
          await mkdir(targetDirname, { recursive: true });
          await copyFile(fileOrDirPath, targetPath);
        }
      })
    );
  }
};
