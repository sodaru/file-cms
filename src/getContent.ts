import { Config } from "./Config";
import { attributeFilter } from "./internal/attributeFilter";
import { fetchContentCached } from "./internal/fetchContentCached";
import { GetContent } from "./types";

export const getContent: GetContent = async (type, slug, attributes) => {
  const rootDir = Config.getRootDir();
  const content = await fetchContentCached(rootDir, type, slug);
  return attributeFilter(content, attributes);
};
