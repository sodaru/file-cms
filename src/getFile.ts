import { join } from "path";
import { Config } from "./Config";
import { readFile } from "fs/promises";
import { FileNotFoundError } from "./errors/FileNotFoundError";

export const getFile = (
  path: string,
  options?: Parameters<typeof readFile>["1"]
) => {
  const rootDir = Config.getRootDir();
  const filePath = join(rootDir, path);
  try {
    return readFile(filePath, options);
  } catch (e) {
    throw new FileNotFoundError(rootDir, path, e);
  }
};
