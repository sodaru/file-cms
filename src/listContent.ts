import path from "path";
import { Config } from "./Config";
import { listDirCached } from "./internal/listDirCached";
import { AnyExpression, Content, ListContent } from "./types";
import { fetchContentCached } from "./internal/fetchContentCached";
import { attributeFilter } from "./internal/attributeFilter";

export const listContent: ListContent = async (
  filters = {},
  attributes = []
) => {
  const rootDir = Config.getRootDir();

  const { type: typeFilter, slug: slugFilter, ...contentFilters } = filters;

  // Level 1, identify the files to fetch
  const filesToFetch: { type: string; slug: string }[] = [];

  const allTypes = await listDirCached(rootDir);
  const typeFilterPredicate = expressionToProvider(typeFilter);
  const typesToFetch = typeFilter
    ? allTypes.filter(typeFilterPredicate)
    : allTypes;

  await Promise.all(
    typesToFetch.map(async type => {
      const allSlugFileNames = await listDirCached(path.join(rootDir, type));
      const allSlugs = allSlugFileNames
        .filter(slug => slug.endsWith(".md"))
        .map(slug => slug.substring(0, slug.length - 3));
      const slugFilterPredicate = expressionToProvider(slugFilter);
      const slugsToFetch = slugFilter
        ? allSlugs.filter(slugFilterPredicate)
        : allSlugs;

      for (const slug of slugsToFetch) {
        filesToFetch.push({ type, slug });
      }
    })
  );

  // level 2, fetch & filter the contents from the selected files
  const allContents = await Promise.all(
    filesToFetch.map(async ({ type, slug }) => {
      return fetchContentCached(rootDir, type, slug);
    })
  );

  const contentPredicate = getContentFilterPredicate(contentFilters);

  const filteredContents = contentPredicate
    ? allContents.filter(contentPredicate)
    : allContents;

  // level 3, filter attributes
  const partialContents = filteredContents.map(content =>
    attributeFilter(content, attributes)
  );

  return partialContents;
};

const getContentFilterPredicate = (filters?: Record<string, AnyExpression>) => {
  const filterKeys = Object.keys(filters);

  if (filterKeys.length > 0) {
    const providers: Record<string, (data: unknown) => boolean> = {};
    filterKeys.forEach(key => {
      providers[key] = expressionToProvider(filters[key]);
    });

    return (content: Content) => {
      for (const key of filterKeys) {
        const provider = providers[key];
        if (!(provider && provider(content[key]))) {
          return false;
        }
      }
      return true;
    };
  }
};

const expressionToProvider = (expression: AnyExpression) => {
  if (expression === undefined) {
    return;
  }
  const fullExpression =
    expression === null ||
    typeof expression == "string" ||
    typeof expression == "number" ||
    typeof expression == "boolean"
      ? { eq: expression }
      : expression;

  const exprName = Object.keys(fullExpression)[0];
  const exprArg = Object.values(fullExpression)[0];
  const provider = Config.getFilterProvider(exprName);
  return (data: unknown) => {
    return provider(data, exprArg);
  };
};
