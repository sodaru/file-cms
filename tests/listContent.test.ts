import path from "path";
import { Config } from "../src/Config";
import { listContent } from "../src/listContent";

const commonRootDir = path.join(__dirname, "sample-data");
describe("listContent Tests", () => {
  test("Filter content by non-existent type", async () => {
    Config.setRootDir(commonRootDir);
    const contents = await listContent({ type: "nonexistentType" });
    expect(contents).toHaveLength(0);
  });
  test("Filter content by type", async () => {
    Config.setRootDir(commonRootDir);
    const contents = await listContent({ type: "blog" });
    expect(contents).toMatchSnapshot();
  });
  test("Filter content by type with no results", async () => {
    Config.setRootDir(commonRootDir);
    const contents = await listContent({ type: "non-existent-type" });
    expect(contents).toEqual([]);
  });

  test("return title of all slugs", async () => {
    Config.setRootDir(commonRootDir);
    const slug = await listContent({}, ["title"]);
    expect(slug).toMatchSnapshot();
  });
  test("Filter content matching in title", async () => {
    Config.setRootDir(commonRootDir);
    const contents = await listContent({ title: "JavaScript" });
    expect(contents).toMatchSnapshot();
  });

  test("Filter content by keyword in title", async () => {
    Config.setRootDir(commonRootDir);
    const contents = await listContent({ title: { contains: "JavaScript" } });
    expect(contents).toMatchSnapshot();
  });

  test("Filter content by date range", async () => {
    Config.setRootDir(commonRootDir);
    const contents = await listContent({
      ReleasedateFrom: "2024-06-01",
      ReleasedateTo: "2024-08-31"
    });
    expect(contents).toMatchSnapshot();
  });

  test("Filter content by keywords", async () => {
    Config.setRootDir(commonRootDir);
    const contents = await listContent({
      tags: { has: "JavaScript" }
    });
    expect(contents).toMatchSnapshot();
  });
  test("Filter content by author", async () => {
    Config.setRootDir(commonRootDir);
    const contents = await listContent({ author: "Alice Smith" });
    expect(contents).toMatchSnapshot();
  });
  test("Check listing all content without filters", async () => {
    Config.setRootDir(commonRootDir);
    const contents = await listContent({});
    expect(contents).toMatchSnapshot();
  });
  test("Check sorting by date", async () => {
    Config.setRootDir(commonRootDir);
    const contents = await listContent({ sortBy: "date", order: "asc" });
    expect(contents).toMatchSnapshot();
  });
  test("Filter content by type and keyword", async () => {
    Config.setRootDir(commonRootDir);
    const contents = await listContent({
      type: "product",
      title: "Noise-Canceling Headphones"
    });
    expect(contents).toMatchSnapshot();
  });
});
