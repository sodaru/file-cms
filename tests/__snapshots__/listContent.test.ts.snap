// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`listContent Tests Check listing all content without filters 1`] = `
Array [
  Object {
    "author": "Alice Smith",
    "content": "# JavaScript Closures

Closures are an important concept in JavaScript, enabling powerful patterns for scoping and memory management.",
    "date": "2024-08-12",
    "slug": "blog-1",
    "tags": Array [
      "JavaScript",
      "Programming",
    ],
    "title": "Understanding JavaScript Closures",
    "type": "blog",
  },
  Object {
    "author": "John Doe",
    "content": "# AI in Healthcare

Artificial intelligence is transforming the healthcare industry by improving diagnosis, treatment, and patient care.",
    "date": "2024-08-18",
    "slug": "blog-2",
    "tags": Array [
      "AI",
      "Healthcare",
    ],
    "title": "The Future of AI in Healthcare",
    "type": "blog",
  },
  Object {
    "author": "Jane Doe",
    "content": "# Mastering CSS

CSS is a crucial part of web development, and improving your skills can significantly enhance your front-end development.",
    "date": "2024-09-01",
    "slug": "blog-3",
    "tags": Array [
      "CSS",
      "Web Development",
    ],
    "title": "How to Improve Your CSS Skills",
    "type": "blog",
  },
  Object {
    "content": "# Python for Data Science

Python is the go-to programming language for data science due to its simplicity and the vast array of libraries available.",
    "slug": "blog-4",
    "type": "blog",
  },
  Object {
    "author": "Sarah Brown",
    "content": "# Productivity Tips

This content includes special characters: @#$%^&().",
    "date": "2024-09-05",
    "slug": "blog-5",
    "tags": Array [
      "Productivity",
      "Developers",
    ],
    "title": "Special @#$% Characters!",
    "type": "blog",
  },
  Object {
    "category": "Electronics",
    "content": "# SuperFast Laptop

A high-performance laptop designed for professionals who need speed and reliability.",
    "price": "$999",
    "releaseDate": "2024-07-10",
    "slug": "product-1",
    "tags": Array [
      "laptop",
      "electronics",
      "featured",
    ],
    "title": "SuperFast Laptop",
    "type": "product",
  },
  Object {
    "category": "Accessories",
    "content": "# Eco-Friendly Water Bottle

Stay hydrated with this eco-friendly, reusable water bottle, perfect for on-the-go lifestyles.",
    "price": "$25",
    "releaseDate": "2024-06-15",
    "slug": "product-2",
    "tags": Array [
      "water bottle",
      "eco-friendly",
    ],
    "title": "Eco-Friendly Water Bottle",
    "type": "product",
  },
  Object {
    "category": "Audio",
    "content": "# Noise-Canceling Headphones

Experience immersive sound with our noise-canceling headphones, ideal for music lovers and travelers.",
    "price": "$199",
    "releaseDate": "2024-08-01",
    "slug": "product-3",
    "tags": Array [
      "headphones",
      "audio",
    ],
    "title": "Noise-Canceling Headphones",
    "type": "product",
  },
  Object {
    "category": "Home Security",
    "content": "# Smart Home Security

Secure your home with our easy-to-install smart security system, controlled via mobile app.",
    "price": "$299",
    "releaseDate": "2024-09-01",
    "slug": "product-4",
    "tags": Array [
      "home security",
      "smart home",
    ],
    "title": "Smart Home Security System",
    "type": "product",
  },
  Object {
    "category": "Furniture",
    "content": "# Electric Standing Desk

Boost your productivity with an adjustable standing desk, designed for comfort and ergonomics.",
    "price": "$450",
    "releaseDate": "2024-07-20",
    "slug": "product-5",
    "tags": Array [
      "desk",
      "office",
      "standing",
    ],
    "title": "Electric Standing Desk using Javascript",
    "type": "product",
  },
  Object {
    "category": "Development",
    "content": "# Website Development

We provide custom website development services tailored to your business needs, from small businesses to large enterprises.",
    "priceRange": "$1000 - $5000",
    "slug": "service-1",
    "tags": Array [
      "web development",
      "services",
    ],
    "title": "Website Development",
    "type": "service",
  },
  Object {
    "category": "Marketing",
    "content": "",
    "priceRange": "$800 - $4000",
    "slug": "service-2",
    "tags": Array [
      "digital marketing",
      "services",
    ],
    "title": "Digital Marketing Strategy",
    "type": "service",
  },
  Object {
    "category": "SEO",
    "content": "# SEO Optimization

Optimize your website for search engines and improve your ranking with our proven SEO services.",
    "priceRange": "$500 - $2000",
    "slug": "service-3",
    "tags": Array [
      "SEO",
      "services",
    ],
    "title": "SEO Optimization using Javascript",
    "type": "service",
  },
  Object {
    "category": "Development",
    "content": "# Mobile App Development

We create powerful and user-friendly mobile applications to help you reach your audience on iOS and Android.",
    "priceRange": "$3000 - $10000",
    "slug": "service-4",
    "tags": Array [
      "mobile development",
      "services",
    ],
    "title": "Mobile App Development",
    "type": "service",
  },
  Object {
    "category": "Hosting",
    "content": "# Cloud Hosting Solutions

Reliable and scalable cloud hosting solutions designed to meet the demands of your growing business.",
    "priceRange": "$150 - $1000/month",
    "slug": "service-5",
    "tags": Array [
      "cloud hosting",
      "services",
    ],
    "title": "Cloud Hosting Solutions",
    "type": "service",
  },
]
`;

exports[`listContent Tests Check sorting by date 1`] = `Array []`;

exports[`listContent Tests Filter content by author 1`] = `
Array [
  Object {
    "author": "Alice Smith",
    "content": "# JavaScript Closures

Closures are an important concept in JavaScript, enabling powerful patterns for scoping and memory management.",
    "date": "2024-08-12",
    "slug": "blog-1",
    "tags": Array [
      "JavaScript",
      "Programming",
    ],
    "title": "Understanding JavaScript Closures",
    "type": "blog",
  },
]
`;

exports[`listContent Tests Filter content by date range 1`] = `Array []`;

exports[`listContent Tests Filter content by keyword in title 1`] = `
Array [
  Object {
    "author": "Alice Smith",
    "content": "# JavaScript Closures

Closures are an important concept in JavaScript, enabling powerful patterns for scoping and memory management.",
    "date": "2024-08-12",
    "slug": "blog-1",
    "tags": Array [
      "JavaScript",
      "Programming",
    ],
    "title": "Understanding JavaScript Closures",
    "type": "blog",
  },
  Object {
    "category": "Furniture",
    "content": "# Electric Standing Desk

Boost your productivity with an adjustable standing desk, designed for comfort and ergonomics.",
    "price": "$450",
    "releaseDate": "2024-07-20",
    "slug": "product-5",
    "tags": Array [
      "desk",
      "office",
      "standing",
    ],
    "title": "Electric Standing Desk using Javascript",
    "type": "product",
  },
  Object {
    "category": "SEO",
    "content": "# SEO Optimization

Optimize your website for search engines and improve your ranking with our proven SEO services.",
    "priceRange": "$500 - $2000",
    "slug": "service-3",
    "tags": Array [
      "SEO",
      "services",
    ],
    "title": "SEO Optimization using Javascript",
    "type": "service",
  },
]
`;

exports[`listContent Tests Filter content by keywords 1`] = `
Array [
  Object {
    "author": "Alice Smith",
    "content": "# JavaScript Closures

Closures are an important concept in JavaScript, enabling powerful patterns for scoping and memory management.",
    "date": "2024-08-12",
    "slug": "blog-1",
    "tags": Array [
      "JavaScript",
      "Programming",
    ],
    "title": "Understanding JavaScript Closures",
    "type": "blog",
  },
]
`;

exports[`listContent Tests Filter content by type 1`] = `
Array [
  Object {
    "author": "Alice Smith",
    "content": "# JavaScript Closures

Closures are an important concept in JavaScript, enabling powerful patterns for scoping and memory management.",
    "date": "2024-08-12",
    "slug": "blog-1",
    "tags": Array [
      "JavaScript",
      "Programming",
    ],
    "title": "Understanding JavaScript Closures",
    "type": "blog",
  },
  Object {
    "author": "John Doe",
    "content": "# AI in Healthcare

Artificial intelligence is transforming the healthcare industry by improving diagnosis, treatment, and patient care.",
    "date": "2024-08-18",
    "slug": "blog-2",
    "tags": Array [
      "AI",
      "Healthcare",
    ],
    "title": "The Future of AI in Healthcare",
    "type": "blog",
  },
  Object {
    "author": "Jane Doe",
    "content": "# Mastering CSS

CSS is a crucial part of web development, and improving your skills can significantly enhance your front-end development.",
    "date": "2024-09-01",
    "slug": "blog-3",
    "tags": Array [
      "CSS",
      "Web Development",
    ],
    "title": "How to Improve Your CSS Skills",
    "type": "blog",
  },
  Object {
    "content": "# Python for Data Science

Python is the go-to programming language for data science due to its simplicity and the vast array of libraries available.",
    "slug": "blog-4",
    "type": "blog",
  },
  Object {
    "author": "Sarah Brown",
    "content": "# Productivity Tips

This content includes special characters: @#$%^&().",
    "date": "2024-09-05",
    "slug": "blog-5",
    "tags": Array [
      "Productivity",
      "Developers",
    ],
    "title": "Special @#$% Characters!",
    "type": "blog",
  },
]
`;

exports[`listContent Tests Filter content by type and keyword 1`] = `
Array [
  Object {
    "category": "Audio",
    "content": "# Noise-Canceling Headphones

Experience immersive sound with our noise-canceling headphones, ideal for music lovers and travelers.",
    "price": "$199",
    "releaseDate": "2024-08-01",
    "slug": "product-3",
    "tags": Array [
      "headphones",
      "audio",
    ],
    "title": "Noise-Canceling Headphones",
    "type": "product",
  },
]
`;

exports[`listContent Tests Filter content matching in title 1`] = `Array []`;

exports[`listContent Tests return title of all slugs 1`] = `
Array [
  Object {
    "title": "Understanding JavaScript Closures",
  },
  Object {
    "title": "The Future of AI in Healthcare",
  },
  Object {
    "title": "How to Improve Your CSS Skills",
  },
  Object {
    "title": undefined,
  },
  Object {
    "title": "Special @#$% Characters!",
  },
  Object {
    "title": "SuperFast Laptop",
  },
  Object {
    "title": "Eco-Friendly Water Bottle",
  },
  Object {
    "title": "Noise-Canceling Headphones",
  },
  Object {
    "title": "Smart Home Security System",
  },
  Object {
    "title": "Electric Standing Desk using Javascript",
  },
  Object {
    "title": "Website Development",
  },
  Object {
    "title": "Digital Marketing Strategy",
  },
  Object {
    "title": "SEO Optimization using Javascript",
  },
  Object {
    "title": "Mobile App Development",
  },
  Object {
    "title": "Cloud Hosting Solutions",
  },
]
`;
